Optimize mapgen and saving/loading for solid rock/open air submaps

Coolthulhu opened the pull request:
Deals with 

> implement special submap type that is homogeneously one terrain(e.g. only air, or only rock)

from #6818

Submaps that are created by filling entire area with one type of terrain (and no furniture, radiation etc.) are marked as uniform. If anything changes (furniture is added, light is brought into any tile etc.), they are no longer uniform.

A quad of uniform submaps will not be saved, because regenerating it is faster than re-reading it.

Optimized filling entire submaps with one terrain type. Before it would do so by invoking `set_ter` in a loop (which then invoked `get_submap_at_grid` each time), now it sets it all in bulk. Does speed up regular mapgen a tiny bit too.

I reverted some minor changes to submap addressing at mapgen. Back to iterating over x and y from iterating over entire vector. This doesn't really affect anything, but will make the switch to z-levels a tiny bit smoother.

This doesn't really affect anything in normal game (maybe a tiny bit faster mapgen/map save speed underground and on experimental z-level buildings), I just wanted to split the z-level updates rather than sending them in bulk.
This change is necessary for z-levels, because without it mapgen, saving and loading would be almost 20 times slower once z-levels are in. I'm sending it before regular z-level update to make transition smooth instead of having forums flooded with "why is my save time 20 times longer now?".

BevapDin commented on src/mapdata.h:
```diff
@@ -395,26 +401,31 @@ struct submap {
+    inline const std::string get_signage(int x, int y) const {
+        furn_id f = frn[x][y];
+        if( furnlist[f].id == "f_sign" ) {
+            return cosmetics[x][y].find("SIGNAGE")->second;
```
You should check the iterator returned by `find`, it's quite possible to only have a `f_sign` there but no entry in the cosmetics map. Mapgen (especially json based) allows this.

BevapDin commented on src/mapdata.h:
```diff
@@ -344,28 +345,33 @@ struct submap {
     }
 
     void update_lum_rem(item const &i, int const x, int const y) {
+        is_uniform = false;
```
AFAIK `lum` is not stored in the save, I understand that `is_uniform` only affects the decision to save the map / not save it, so a different `lum` value should not change `is_uniform`. Adding a (light emitting) item to the submap will trigger `is_uniform = false;` anyway.

BevapDin commented:
Nice work. More const-correctness is always good.

You've missed several places in map.cpp that access the submap data directly, e.g. the whole vehicles (the submap must be stored if a vehicle moves onto it to store the vehicle as a well), also `map::ter_set`, `map::add_item_at`, changing fields through the field object returned by `map::get_field`.

Some of those places could be change to use setter/getter function in the submap. Btw. function declarations inside the class declaration (e.g. `get_radiation` in submap) are automatically inline, no need to specify that explicitly.

Also: maybe it is useful to check the preexisting terrain/furniture/etc. against the new one and skip the setting if it would be unchanged (e.g. `set_ter(t_rock)` makes nothing if the terrain at the target is already `t_rock`). But that might create additional processing overhead.

Coolthulhu commented:
> You've missed several places in map.cpp that access the submap data directly, e.g. the whole vehicles (the submap must be stored if a vehicle moves onto it to store the vehicle as a well), also map::ter_set, map::add_item_at, changing fields through the field object returned by map::get_field.

I'll fix `ter_set` and `add_item_at` in a second (I fixed them in my main z-level branch, just forgot to copy them here). I'll leave fields and vehicles as is for now, because they shouldn't really come in contact with uniform submaps - fields would need to be created first (violating uniformness), vehicles in uniform maps act really stupid anyway (only open air and solid rock maps are uniform at the moment).
I'll fix those two later in 3D field and vehicle updates.

I'll change all the direct array accesses to getters/setters, though probably not in this PR. There's a comment in `mapdata.h` indicating that all the fields like `ter[x][y]` should be private, so this change is desirable anyway.
For this one, I changed most of the direct array writes. Left `mapgen::rotate` with direct writes and added `is_uniform = false` instead, not to slow it down.

> Also: maybe it is useful to check the preexisting terrain/furniture/etc. against the new one and skip the setting if it would be unchanged

For now I'd leave it as is not to slow down mapgen. Later on it uniformness could be re-calculated at save, furniture stored as map instead of as an array etc., but at the moment most 100% solid terrain submaps are never modified.
When furniture starts falling across z-levels, it would be nice to have uniformness re-applied.

kevingranade commented:
Only thing that jumped out at me reading the description and discussion (on
bus so can't really review the code) is that iterating array-wise over
submaps instead of coordinate-wise over maps increased cache coherency
massively in a recent PR (about 2x speedup), so we want to be careful about
reversing that.

Coolthulhu commented:
The recent speedup is probably related to lightmap calculation. I didn't touch this one here and the changes are to non-performance-critical sections of the code.

I'll keep that in mind, though. I may have to change the way I'm addressing nonants in my z-level branch to make it more cache-friendly.

Zireael07 commented:
@Coolthulhu: Great work!
